fauna/client.py

```diff
@@ -256,85 +259,118 @@ def _query(
             headers = response.headers()
             status_code = response.status_code()
 
+            self._check_protocol(response_json, status_code)
+
             if status_code > 399:
-                Client._handle_error(response_json, status_code)
+                self._handle_error(response_json, status_code)
 
             if "txn_ts" in response_json:
                 self.set_last_txn_ts(int(response_json["txn_ts"]))
 
             return QueryResponse(response_json, headers, status_code)
 
-    @staticmethod
-    def _handle_error(response_json: Any, status_code: int):
-        if "error" not in response_json:
+    def _check_protocol(self, response_json: Any, status_code):
+        # TODO: Logic to validate wire protocol belongs elsewhere.
+        should_raise = False
+
+        # check for QuerySuccess
+        if status_code <= 399 and "data" not in response_json:
+            should_raise = True
+
+        # check for QueryFailure
+        if status_code > 399:
+            if "error" not in response_json:
+                should_raise = True
+            else:
+                e = response_json["error"]
+                if "code" not in e or "message" not in e:
+                    should_raise = True
+
+        if should_raise:
             raise ProtocolError(
                 status_code,
                 "Unexpected response",
-                response_json,
+                f"Response is in an unknown format: \n{response_json}",
             )
 
-        err = ServiceError(
-            status_code,
-            response_json["error"]["code"],
-            response_json["error"]["message"],
-            response_json["summary"] if "summary" in response_json else "",
-        )
+    def _handle_error(self, response_json: Any, status_code: int):
+        err = response_json["error"]
+        code = err["code"]
+        message = err["message"]
+        summary = response_json["summary"] if "summary" in response_json else ""
+        constraint_failures: Optional[List[ConstraintFailure]] = None
+
+        if "constraint_failures" in err:
+            constraint_failures = [
+                ConstraintFailure(
+                    message=cf["message"],
+                    name=cf["name"] if "name" in cf else None,
+                    paths=cf["paths"] if "paths" in cf else None,
+                ) for cf in err["constraint_failures"]
+            ]
+
         if status_code == 400:
-            if err.code is not None:
+            if code == "invalid_query":
                 raise QueryCheckError(
-                    err.status_code,
-                    err.code,
-                    err.message,
-                    err.summary,
+                    status_code,
+                    code,
+                    message,
+                    summary,
+                )
+            else:
+                raise QueryRuntimeError(
```

Comment:
out of scope for this PR.

But was talking with Paul about whether we should get rid of the QueryRuntime / QueryCheck distinction in the drivers.

There really the same semantically for the user - my query is no good.

Would also decouple us from having to map codes to error types.

Putting here for discussion.

Reply:
Yeah, I'm down for consolidating this to QueryError. If we see users asking for special types, we can always inherit from QueryError at that point